app/controllers/application_controller.rb

```diff
@@ -1,10 +1,172 @@
 
 class ApplicationController < Sinatra::Base
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+  	set :session_secret, "secret"
   end
 
-end
\ No newline at end of file
+
+ #Home
+  get '/' do 
+  	erb :'index'
+  end
+
+
+#Sign Up
+  get '/signup' do
+  	if Helpers.is_logged_in?(session)
+  		redirect '/tweets'
+  	else
+  		erb :'users/signup'
+  	end
+  end
+
+  post '/signup' do 
+  	@user = User.create(:username => params[:username], :email => params[:email], :password => params[:password])
+  	@user.save
+  	session[:id] = @user[:id]
+  	if @user.save && !params[:username].empty? && !params[:email].empty? && !params[:password].empty?
+  		redirect '/tweets'
+  	else
+  		redirect '/signup'
+  	end
+  end
+
+#Log in
+  get '/login' do
+  	if Helpers.is_logged_in?(session)
+  		redirect '/tweets'
+  	else
+		erb :'users/login'
+	end
+  end
+
+  post '/login' do 
+ 	@user = User.find_by(:username => params[:username])
+ 	if @user && @user.authenticate(params[:password])
+  		session[:id] = @user[:id]
+		redirect '/tweets'
+	else
+		redirect '/login'
+	end
+  end
+
+  #Log Out
+  get '/logout' do 
+  	session.clear
+  	redirect '/login'
+  end
+
+  #Home Page - tweets feed
+  get '/tweets' do
+  	if Helpers.is_logged_in?(session)
+  		@user = User.find_by_id(session[:id])
+  		erb :'tweets/tweets'
+  	else
+  		redirect '/login'
+  	end
+  end
+
+  #Show single user tweets
+  get '/users/:slug' do 
+  	@user = User.find_by_slug(params[:slug])
+  	erb :'users/tweets'
+  end
+
+  #TWEETS
+  #New
+  get '/tweets/new' do 
+  	if Helpers.is_logged_in?(session)
+	  	@user = User.find_by_id(session[:id])
+  		erb :'tweets/new'
+  	else
+  		redirect '/login'
+  	end
+  end
+
+  post '/tweets/new' do 
+  	if params[:content] != ""
+  		@tweet = Tweet.create(:content => params[:content])
+  		@user = User.find_by_id(session[:id])
+  		@user.tweets << @tweet
+  		redirect "/tweets/#{@tweet.id}"
+  	else
+  		redirect '/tweets/new'
+  	end
+  end	
+
+  #Show
+  get '/tweets/:id' do
+  	if Helpers.is_logged_in?(session)
+  		@tweet = Tweet.find_by(:id => params[:id])
+  		erb :'tweets/show'
+  	else
+  		redirect '/login'
+  	end
+  end
+
+  #Edit
+  get '/tweets/:id/edit' do 
+  	if Helpers.is_logged_in?(session)
+  		@tweet = Tweet.find_by(:id => params[:id])
+  		erb :'tweets/edit'
+  	else
+  		redirect '/login'
+  	end
+  end
+
+  patch '/tweets/:id/edit' do
+  	@tweet = Tweet.find_by(:id => params[:id])
+    @user = User.find_by_id(session[:id])
+    redirect "/tweets" if @user.id != @tweet.user_id
+  	if params[:content] != ""
+  		#binding.pry
+  		@tweet.content = params[:content]
+  		@tweet.save
+  		redirect "/tweets/#{@tweet.id}"
+  	else
+  		redirect "tweets/#{@tweet.id}/edit"
+  	end
+  end
+
+  #Delete
+  delete '/tweets/:id/delete' do 
+  	@user = User.find_by_id(session[:id])
+  	@tweet = Tweet.find_by(:id => params[:id])
+  	Tweet.delete(params[:id]) if @user.id == @tweet.user_id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.